src/portable/synopsys/dwc2/dcd_dwc2.c

```diff
@@ -67,6 +67,16 @@ static bool _sof_en;
 //--------------------------------------------------------------------
 // DMA
 //--------------------------------------------------------------------
+// When DMA requires cache synchronization for memory
```

Comment:
@hathach @HiFiPhile 
Hi,
may I ask to briefly check the changes in `dcd_dwc2.c` and share your opinion? 

I've implemented the cache synchronization for Device and tested it on esp32p4, port1 (HS).
I've tried to implement the feature and: 
-  Not to break the dcd_dwc2 for other controllers (where cache calls are not necessary or not implemented)
- Put the calls in the logically important places in the code (to be clear why do we need sync cache/mem)
- Use the sections and memory alignment as much as possible. 

But on P4 almost all the major memory which is visible for DMA is behind L1-Cache, so ... we need everything: section + alignment + cache sync. 

I might have miss something, because I didn't test the Host at all (if there will be any suggestions, how I can do that quickly and easy, please let me know). 

Anyway, will appreciate any feedback from your side. Thanks.

Reply:
Thank you for your effort, I only have a preliminary datasheet in which it mentioned cacheable and non-cacheable access for L1 cache, maybe it supports configuring memory attribute like Cortex-M MPU ?

I think this is not necessary:
```c
    if(dma_device_enabled(dwc2)) {
      dep->doepdma = (uintptr_t)xfer->buffer;
      dsync_c2m(xfer->buffer, total_bytes);
```

Also we need to ensure `_setup_packet` is aligned to cache line and **has a size of cache line**

Reply:
@HiFiPhile yes, you are right, not necessary. Thanks for checking. 

Currently, it is with the following alignment `TU_ATTR_ALIGNED(4)` and maybe I can change it to `CFG_TUD_MEM_ALIGN`. 
While usually: `#define CFG_TUD_MEM_ALIGN        __attribute__ ((aligned(4)))` the change shouldn't break anything. 

Thanks.

Reply:
... or it will. 

Do you know any other scenarios, when we need to provide `CFG_TUD_MEM_ALIGN` which is different from   `__attribute__ ((aligned(4)))` ?

Reply:
Hi @tore-espressif may I ask a question, does ESP32P4 support memory management to mark a RAM section as non-cacheable for DMA use, similar to Cortex-M MPU ?

> Do you know any other scenarios, when we need to provide CFG_TUD_MEM_ALIGN which is different from __attribute__ ((aligned(4))) ?

For Cortex-M if cache invalidate/clean is used instead of making non-cacheable section by MPU, `_setup_packet` has to be aligned to 32 or 64 bytes depending the core, also it's size has to be grown to 32/64 bytes matching cache line size.

Otherwise other variables around will be erased on cache invalidate, it's explained in page 10: https://ww1.microchip.com/downloads/en/DeviceDoc/Managing-Cache-Coherency-on-Cortex-M7-Based-MCUs-DS90003195A.pdf

Reply:
@HiFiPhile 

> I think this is not necessary:

Unfortunately, it is necessary. At least in current implementation when the class driver is using class specific requests. 
For example, in `msc class`, when we schedule the `cbw`, we need to sync c2m before getting the IRQ.
Otherwise, the data will be lost. 

I will think, where to move the `c2m` call in case of OUT scheduling not to confuse developers that much.

Reply:
@HiFiPhile ,

Thanks for the reason to double check it!

It is not necessary. What I had before - was a collateral damage because of this TODO: https://github.com/hathach/tinyusb/blob/0569188aed17d19d5954f1e1b911e989814ec93d/src/class/msc/msc_device.c#L58

If mscd `cbw` and `csw` buffers will be not the members of `_mscd_itf` and cache line aligned - no problem and we don't need the initial line to synchronize cache.
When I synchronize it before, I damaged the `ep_out` member in memory and the first SCSI request failed.  

Thanks again, I can fix the mscd driver with optimizing the `cbw` and `csw` buffers and make a PR with changes. 

In general, refuse to use cache sync in here maybe a good idea. In general and when it is possible. 
I will check if it is possible on P4.

Reply:
@HiFiPhile @roma-jam The ESP32-P4 uses RISC-V cores, so while there are differences, the concept is similar. Specifically, the ESP32-P4 implements RISC-V Physical Memory Attribution (PMA), which functions similarly to ARM's Memory Protection Unit (MPU).

The PMA imposes certain constraints on memory sections, such as requiring sizes to be powers of 2 and start addresses to align with the section size. Addressing these limitations in TinyUSB would require some refactoring, and certain changes might need to be platform-specific.

If TinyUSB has previously resolved similar issues on ARM by placing buffers in non-cachable memory, adopting the same approach for the ESP32-P4 might be the most straightforward solution.

Reply:
Thank you @tore-espressif, so it's basically:
- Create the section in linker with correct size /alignment
- Configure PMA
- Use `CFG_TUSB_MEM_SECTION` to put buffers in the section

?

Reply:
It is even simpler, we have 1 physical memory mapped to 2 virtual addresses, one is cachable (default) the other one is non-cachable. So we just use the non-cachable address. @roma-jam will test this soon.

Unfortunately, the reference manual hasn't been made public yet

Reply:
Hi everyone, 

I have implemented cache bypassing for dcd_dwc2, usbd_control and mscd class to verify it.
The draft could be found here: https://github.com/espressif/tinyusb/pull/38

But it gives us the picture of what should be done in all device classes, because all the buffers, which could be transferred via DMA and used at the higher layer should be noncache-able...